syncstorage/views/__init__.py

```diff
@@ -319,6 +342,84 @@ def post_collection(request):
     return res
 
 
+def post_collection_transaction(request):
+    storage = request.validated["storage"]
+    userid = request.validated["userid"]
+    collection = request.validated["collection"]
+    bsos = request.validated["bsos"]
+    invalid_bsos = request.validated["invalid_bsos"]
+    batch = request.validated["batch"]
+    commit = request.validated["commit"]
+
+    # The "batch" key is set only on a multi-POST batch request prior to a
+    # commit.  The "modified" key is only set upon a successful commit.
+    # The two flags are mutually exclusive.
+    # Any failures at all mean cancelling the transaction completely.
+    res = {'success': [], 'failed': {}}
+
+    # If there are any parsing failures, we won't even start a transaction.
+    if len(invalid_bsos):
+        for (id, error) in invalid_bsos.iteritems():
+            res["failed"][id] = error
+        # if batch and batch is not True:
+        #     storage.delete_transaction(batch)
+        return res
+
+    try:
+        if batch is True:
+            try:
+                batch = storage.create_transaction(userid, collection)
+            except ConflictError, e:
+                # ConflictError here means a client is spamming requests,
+                # I think.
+                logger.error('Collision in transaction creation!')
+                logger.error(e)
+                raise HTTPServiceUnavailable('Too many transactions opened')
+            except Exception, e:
+                logger.error('Could not create transaction')
+                logger.error(e)
+                raise
+
+        try:
+            ts = storage.append_items_to_transaction(batch, userid,
+                                                     collection, bsos)
+        except InvalidTransaction:
+            raise HTTPBadRequest("Invalid or expired transaction ID")
+        except ConflictError:
+            raise
+        except Exception, e:
+            logger.error('Could not append to transaction'
+                         '("{0}")'.format(batch))
+            logger.error(str(e))
+            for bso in bsos:
+                res["failed"][bso["id"]] = "db error"
+            raise
+
+        if commit:
+            try:
+                storage.commit_transaction(batch, userid, collection)
+                res['modified'] = ts
```

Comment:
IIRC, we want all items in the transaction to get the timestamp from the time of commit.  It may be slightly more reliable to use `ts = store.commit_transaction()` here rather than using the timestamp from the previous `append_items_to_transaction`.